Remotion/Validation/Core/Attributes/Validation/NotEmptyValidationAttribute.cs

```diff
@@ -38,20 +61,34 @@ protected override IEnumerable<IPropertyValidator> GetValidators (IPropertyInfor
       ArgumentUtility.CheckNotNull("property", property);
       ArgumentUtility.CheckNotNull("validationMessageFactory", validationMessageFactory);
 
-      NotEmptyValidator validator;
-      if (string.IsNullOrEmpty(ErrorMessage))
-      {
-        validator = PropertyValidatorFactory.Create(
-            property,
-            parameters => new NotEmptyValidator(parameters.ValidationMessage),
-            validationMessageFactory);
-      }
+      IPropertyValidator validator;
+      if(typeof(string).IsAssignableFrom(property.PropertyType))
+        validator = CreateValidator(msg => new NotEmptyStringValidator(msg));
+      else if (typeof(byte[]).IsAssignableFrom(property.PropertyType))
+        validator = CreateValidator(msg => new NotEmptyBinaryValidator(msg));
+      else if(typeof(ICollection).IsAssignableFrom(property.PropertyType))
+        validator = CreateValidator(msg => new NotEmptyCollectionValidator(msg));
+      else if (typeof(IReadOnlyCollection<object>).IsAssignableFrom(property.PropertyType))
+        validator = CreateValidator(msg => new NotEmptyCollectionValidator(msg));
+      else if(property.PropertyType.CanAscribeTo(typeof(ICollection<>)))
+        validator = CreateValidator(msg => new NotEmptyCollectionValidator(msg));
       else
-      {
-        validator = new NotEmptyValidator(new InvariantValidationMessage(ErrorMessage));
-      }
+        throw new NotSupportedException($"{nameof(NotEmptyValidationAttribute)} is not supported for properties of type '{property.PropertyType.FullName}'.");
 
       return EnumerableUtility.Singleton(validator);
+
+      IPropertyValidator CreateValidator (Func<ValidationMessage, IPropertyValidator> factory)
+      {
+        if (string.IsNullOrEmpty(ErrorMessage))
+        {
+          return PropertyValidatorFactory.Create(
+              property,
+              parameters => factory(parameters.ValidationMessage),
+              validationMessageFactory);
+        }
+
+        return factory(new InvariantValidationMessage(ErrorMessage));
```

Comment:
Code style: we try to do balanced IFs when the code should be symmetrical: only guard/exit conditions get their stand-alone IF. Business logic that has two (or more) equal branches should also be set in an IF/ELSE to make this obvious.